Sites Dashboard: Use `react-query` to fetch `SiteExcerptData`

danielbachhuber opened the pull request:
Alternative to #65593, based on feedback from @tyxla https://github.com/Automattic/wp-calypso/pull/65593#pullrequestreview-1040027902
See #65490

## Proposed Changes

Introduces a `SiteExcerptData` concept, powered by `react-query`, to display a limited set of site data on the Sites Dashboard for better performance. See analysis in pdKhl6-v6-p2

**This pull request is not yet complete.** I'm looking for: feedback on the overall approach, as well as opinions on what "complete" looks like from a technical perspective (tests, etc.)

Consider this an opportunity to educate a PHP hacker!

## Testing Instructions

1. Navigate to `/sites-dashboard`.
2. Clear the `calypso_store` in IndexDB.
3. Refresh the page.
4. Observe your list of sites load after the API request completes.
5. Observe new `calypso_store` entries for `query-state-<num>` and `redux-state-<num>`.
6. Delete the `query-state-<num>` entry.
7. Refresh the page.
8. Observe your list of sites appear without delay.
9. Observe the `query-state-<num>` populated from Redux data (each site will have fields that aren't present in the excerpts subset of data).

## Pre-merge Checklist

Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.

- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in [Simple](P9HQHe-k8-p2-p2), [Atomic](P9HQHe-jW-p2-p2), and [self-hosted Jetpack sites](PCYsg-g6b-p2-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Are we memoizing when appropriate (for expensive computations)? More info in [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md) and [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors)
- [ ] Have we sent any new strings [for translation](PCYsg-1vr-p2-p2) ASAP?

Related to #

danielbachhuber committed:
Sites Dashboard: Use `react-query` to fetch `SiteExcerptData`

danielbachhuber commented on client/sites-dashboard/use-sites-data-query.ts:
```diff
@@ -0,0 +1,45 @@
+};
+
+export const useSitesDataQuery = () =>
+	useQuery( [ 'sites-dashboard-sites-data' ], fetchSites, {
```
Looking for input on a couple of things here:

1. What's the best way to return `useSelector( getSites )` _only_ when `state.sites.items !== null` ?
    
    I tried passing `useSelector( getSites )` as `initialData` but it caused the `react-query` cache to be null when the state was initially null.
    
    Furthermore, using `useSelector()` inside of `useSitesDataQuery` seems to cause the function to be called often.
    
2. How can I clear this `react-query` cache when the Redux store is updated?

zaguiini replied:
Hey Daniel, for #1 that's what I came up with:

```diff
diff --git a/client/sites-dashboard/components/sites-dashboard.tsx b/client/sites-dashboard/components/sites-dashboard.tsx
index 649b2c9daf..45e192aa46 100644
--- a/client/sites-dashboard/components/sites-dashboard.tsx
+++ b/client/sites-dashboard/components/sites-dashboard.tsx
@@ -59,9 +59,9 @@ const DashboardHeading = styled.h1`
 
 export function SitesDashboard( { launchStatus }: SitesDashboardProps ) {
 	const { __ } = useI18n();
-	const { data: sites } = useSitesDataQuery();
+	const { data: sites, isLoading } = useSitesDataQuery();
 
-	if ( ! sites ) {
+	if ( ! sites || isLoading ) {
 		return null;
 	}
 
diff --git a/client/sites-dashboard/use-sites-data-query.ts b/client/sites-dashboard/use-sites-data-query.ts
index 50ed22c7db..256842faa2 100644
--- a/client/sites-dashboard/use-sites-data-query.ts
+++ b/client/sites-dashboard/use-sites-data-query.ts
@@ -1,7 +1,10 @@
 import config from '@automattic/calypso-config';
 import { useQuery } from 'react-query';
+import { useSelector } from 'react-redux';
 import wpcom from 'calypso/lib/wp';
+import getSites from 'calypso/state/selectors/get-sites';
 import { SiteData, SiteDataOptions } from 'calypso/state/ui/selectors/site-data';
+import { notNullish } from './util';
 
 // Performance-optimized request for lists of sites.
 // Don't add more fields because you will make the request slower.
@@ -21,7 +24,7 @@ export const SITE_EXCERPT_REQUEST_FIELDS = [
 export const SITE_EXCERPT_REQUEST_OPTIONS = [ 'is_wpforteams_site' ] as const;
 
 export type SiteExcerptData = Pick< SiteData, typeof SITE_EXCERPT_REQUEST_FIELDS[ number ] > & {
-	options: Pick< SiteDataOptions, typeof SITE_EXCERPT_REQUEST_OPTIONS[ number ] >;
+	options?: Pick< SiteDataOptions, typeof SITE_EXCERPT_REQUEST_OPTIONS[ number ] >;
 };
 
 const fetchSites = (): Promise< { sites: SiteExcerptData[] } > => {
@@ -37,9 +40,18 @@ const fetchSites = (): Promise< { sites: SiteExcerptData[] } > => {
 	} );
 };
 
-export const useSitesDataQuery = () =>
-	useQuery( [ 'sites-dashboard-sites-data' ], fetchSites, {
-		enabled: true,
+export const useSitesDataQuery = () => {
+	const sites = useSelector( getSites );
+
+	const reduxData = {
+		sites: sites?.filter( notNullish ),
+	};
+
+	const result = useQuery( [ 'sites-dashboard-sites-data' ], fetchSites, {
 		staleTime: 1000 * 60 * 5, // 5 minutes
-		select: ( data ) => data?.sites ?? [],
+		select: ( data ) => data?.sites,
+		placeholderData: reduxData,
 	} );
+
+	return result;
+};
```

It should be pretty straightforward, but let me know if you run into questions.

For #2: what exactly do you want to achieve?

danielbachhuber replied:
> It should be pretty straightforward, but let me know if you run into questions.

Thanks for the example, @zaguiini !

Incorporating some learnings in our Slack chat, here's what I ended up with: 61a4fa18535a52561d86aebe992b185651ca14e3

Notably, I'm using `useStore().getState().sites.items` because I only want to access the _initial_ state of the store. When I use `useSelector( getSites )`, `useSitesDataQuery` is called every time the store updates:

![image](https://user-images.githubusercontent.com/36432/179603260-eb4c974d-6f35-4ead-a8da-b0bf1ce4f0d2.png)

> For 2: what exactly do you want to achieve?

When `useStore().getState().sites.items` is updated, I want to make sure the `react-query` cache is cleared.

p-jackson replied:
😕 I'm a little confused because it seems contradictory. We're using `useStore().getState().sites.items` because we only want the initial state. But then you ask how do I know when it's updated, so it sounds like you _do_ want to do something every time it changes...

It seems to be working fine without clearing the `react-query` cache, so I think I'll merge with this part of it working as is.

danielbachhuber committed:
Fix TypeScript warning

zaguiini commented on client/sites-dashboard/components/searchable-sites-table.tsx:
```diff
@@ -21,7 +21,7 @@ export function SearchableSitesTable( { sites }: SearchableSitesTableProps ) {
 		}
 
-		return searchCollection( sites, term.toLowerCase(), [ 'URL', 'domain', 'name', 'slug' ] );
+		return searchCollection( sites, term.toLowerCase(), [ 'URL', 'name', 'slug' ] );
```
Why did we remove `domain`? 🤔

danielbachhuber replied:
> Why did we remove `domain`? 🤔

It isn't present in `SiteExcerptData`, adds ~0.2 seconds to my user's request, and seems to duplicate `URL`.

Is there a reason it should be included?

vykes-mac commented on client/sites-dashboard/use-sites-data-query.ts:
```diff
@@ -0,0 +1,45 @@
+
+export const SITE_EXCERPT_REQUEST_OPTIONS = [ 'is_wpforteams_site' ] as const;
+
+export type SiteExcerptData = Pick< SiteData, typeof SITE_EXCERPT_REQUEST_FIELDS[ number ] > & {
```
cool trick. Was gonna suggest this, we now change fields one place to add/remove fields for our excerpt.

p-jackson replied:
Yeah nice 😎 I haven't indexed into an array type using the `[ number ]` trick before 👍

danielbachhuber committed:
Merge branch 'trunk' into improve/sites-dashboard-use-react-query

danielbachhuber committed:
Use initial state of Calypso store (if it exists) to prime cache

Avoids `useSelector()` because we don't want to call `useSitesDataQuery`
over and over as the sites store changes.

danielbachhuber commented:
@p-jackson If you like the overall approach, feel free to make any minor tweaks you see and 🚢

vykes-mac commented on client/sites-dashboard/use-sites-data-query.ts:
```diff
@@ -0,0 +1,60 @@
+	const sites = useStore().getState().sites.items;
+
+	let reduxData = undefined;
+	if ( sites && Object.values( sites ) ) {
```
was wondering if might structure this slightly different to avoid calling `Object.values` twice: something like 

```
const data = sites && Object.values( sites );
	if ( data ) {
		reduxData = {
			sites: data.filter( notNullish ),
		};
	}
```

It's negligible but just a suggestion.

p-jackson committed:
Move hook to the shared folder for data querying

p-jackson committed:
Use getSites instead of Object.values to get the initial data

p-jackson committed:
Remove unnecessary null filtering from SitesDashboard

p-jackson commented:
I've updated the PR:
- I've moved the hook to `/client/data`, which is where the other `react-query` hooks for generic Calypso data live
- Using the `getSites` selector again, since the knowledge about the structure of the Redux store is supposed to live in reducers (i.e. we're not supposed to know that `sites.items` exists). It also means that we get the primary site appearing at the top of the list again. Because we're using the raw selector without `useSelector` it means we're not getting updates.
- The `initialData` option lets you pass a callback if you need to do som non-trivial logic to create the initial data
- Even though we're providing a placeholder for the site excerpt's data, `react-query`'s types still insist that `data` might be `undefined`, so I've had to provide a default value here as well: https://github.com/Automattic/wp-calypso/blob/616caa0295d84dc88c8e3d8569c59c35aeec694a/client/sites-dashboard/components/sites-dashboard.tsx#L61
